### PR TITLE
Handle 429 and prevent 301 in `gp_downloadmetadata`

### PR DESCRIPTION
**Pull request by mokagio**
During the WordPress 20.6 finalization, we noticed that `gp_downloadmetadata` didn't offer to retry the GlotPress request when receiving a 429 (see details in [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/19267#discussion_r960523291) and [this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/19275/commits/7c3289fcc7fe0b493aa9a49696bb3ba85a85646a))

This PR adds a crude retry mechanism to the action.

image

While working on this, I noticed that the URL used was lacking the `/`, consistently resulting in a 301 from the API. I addressed that, too.

`gp_downloadmetadata` is set to be removed in favor of a better alternative as soon as we'll have time for it. I tried to write new code that was good, but kept the refactoring and improvements to a minimum.

### Testing

I tested this by pointing my  WordPress iOS to my local copy of the repo.

```diff
--- a/fastlane/Pluginfile
+++ b/fastlane/Pluginfile
@@ -8,7 +8,8 @@ end

 # This comment avoids typing to switch to a development version for testing.
 # gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit', branch: 'tru>
-gem 'fastlane-plugin-wpmreleasetoolkit', '~> 5.1'
+# gem 'fastlane-plugin-wpmreleasetoolkit', '~> 5.1'
+gem 'fastlane-plugin-wpmreleasetoolkit', path: '../../release-toolkit'
```

I then run `bundle exec fastlane download_wordpress_localized_app_store_metadata` till I got a 429.

### CODE REVIEW

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/common/gp_downloadmetadata_action.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/common/gp_downloadmetadata_action.rb:22]: Converted the `puts` to `UI.message` for more consistent logs.

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb:119]: I decided to make this `error` so it would print red to mark the difference in behavior.

image
_(The `301` is because I hadn't updated the URLs yet)_

I can see how `error`/red might be misleading though, and look like a wholesale failure. If you think that's too confusing, I can update to a `message`:

```suggestion
            UI.message("Abandoning `#{locale}` download as requested.")
```

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (crazytonyli) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb:18]: Just a thought, not related to this PR: it might be good to use faraday which has features like following redirects and retry with an interval (which could be a replacement for handling 429 error?).

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (crazytonyli) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb]: Should we stop the execution when encounter this error?

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (mokagio) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb:18]: Yeah... I don't know about [faraday](https://lostisland.github.io/faraday/usage/) or other Ruby networking clients, but given the number of network requests we do with this toolkit, having a feature-rich way to do requests would be handy.

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (mokagio) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb]: Good question 🤔 I'm unsure. Is an error on one locale worth blocking the entire process? From one point of view it might be, so that we can go on and fix it. From another, HTTP comes with occasional transport errors and it would be good to update _some_ metadata at least.

Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb]: Middle-ground: ask the user 😄 

```suggestion
          message = "Received unexpected #{response.code} from request to URI #{response.uri}."
          UI.abort_with_message!(message) unless UI.confirm("#{message} Continue anyway?")
```

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb:18]: iinm Farady is also the network client that fastlane itself uses, so it should already be part of our ruby dependencies anyway.

This is one of those items part of a long list of little things I'd love to improve / refactor to clean up the release-toolkit in many areas and modernize it… but never got the bandwidth to 😅

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb:119]: Given the error message is pretty clear and it's in reaction of us explicitly replying `n` in the prompt on the previous line, I vote for keeping it red with `UI.error` 👍

**Review by AliSoftware — DISMISSED** [review]
You're missing a `CHANGELOG.md` entry, so I'm blocking the merge to avoid accidentally landing this without it 😉 
But apart from that, the rest LGTM 👍  So should be good to ship once you add that.

**Review by crazytonyli — APPROVED** [review]

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb
Comment (mokagio) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb]: Nice, so there's less chances it'll be missed, too.

Comment (mokagio) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/metadata_download_helper.rb]: Just a note to say that I "tested" this before merging:

image

File: CHANGELOG.md
Comment (AliSoftware) [inline CHANGELOG.md:9]: @mokagio I'm not sure this is really a breaking change — as in, that won't require any change to client repos adopting the new release-toolkit to update their call sites or config or whatnot, will it?

I think it's not worth a major version bump on the next release, while a New Feature and minor bump would make more sense for this change 🙃

Comment (mokagio) [reply CHANGELOG.md:9]: Yep, not at all a breaking change. That was an oversight which I realized only after merging this. It's fixed in 39808dab6bace7ed77e403c36494422303127647.

Comment (AliSoftware) [reply CHANGELOG.md:9]: Oh, perfect, thanks for following-up 👍

**Pull request merged**
